Tutorial - Setting up Ruby on Rails analytics, feature flags, and more (#5209)
PostHog/posthog.com: contents/tutorials/ruby-on-rails-analytics.md

---
title: How to set up Ruby on Rails analytics, feature flags and more
date: 2023-02-02
author: ["ian-vanagas"]
showTitle: true
sidebar: Docs
featuredImage: ../images/tutorials/banners/ruby-on-rails-analytics.png
topics: ["configuration", "feature flags", "persons", "events"]
---

Ruby on Rails is a popular fullstack web framework used by companies like Shopify, GitHub, Twitch, and more.

> In this tutorial, we build a basic Ruby on Rails app, add PostHog to it, and then set up some of the key features of PostHog like custom event capture, user identification, and feature flags.

Reviewer: ```suggestion
In this tutorial, we build a basic [Ruby on Rails](https://rubyonrails.org/) app, add PostHog to it, and then set up some of the key features of PostHog, such as custom event capture, user identification, and feature flags.
```

> > Already know how to build a Ruby on Rails app? Skip to the [PostHog integration](#2-integrating-the-posthog-snippet-and-ruby-library).

Reviewer: ```suggestion
> Already know how to build a Ruby on Rails app? Skip to the [PostHog integration step](#2-integrating-the-posthog-snippet-and-ruby-library).
```

## 1. Creating a basic Ruby on Rails app

First, install Ruby. I used `rbenv` and a `~/.bash_profile` to install version `3.1.3` but there are many ways you can do it. After installing the right version of Ruby, install Rails.

```bash
gem install rails
```

Once done, we can start our blog by running the `rails new` command. going into the blog folder, and starting the server.

```bash
rails new blog
cd blog
bin/rails server
```

Going to `localhost:3000` pulls up a site that looks like this:

![App](../images/tutorials/ruby-on-rails-analytics/app.png)

> A more detailed version of this Ruby app building process is in the [Getting Started with Rails](https://guides.rubyonrails.org/getting_started.html) guide.

### Setting up the articles routes

Rails is nice because it has a bunch of commands you can use to generate parts of the app. Use `generate controller` to create the controller for our articles:

```bash
bin/rails generate controller Articles index --skip-routes
```

> We skip the routes because we set those up manually. In `config/routes.rb`, add the details to our route. The `resources` method lets us define a bunch of routes we want all a once:

Reviewer: ```suggestion
We skip the routes because we will set those up manually. In `config/routes.rb`, add details to the route. The `resources` method lets us define a bunch of routes we want all a once:
```

```html
<h1><%= @article.title %></h1>

<p><%= @article.body %></p>
```

### Creating new articles

The last thing to do is add the ability to create a new article. First, in `app/controllers/articles_controller.rb`, add a `new` and `create` function. We want to define the article params in another function to ensure title, body, and author are all included in the request as well:

```ruby
# ...
def new
    @article = Article.new
  end

  def create
    @article = Article.new(article_params)

    if @article.save
      redirect_to @article
    else
      render :new, status: :unprocessable_entity
    end
  end
	
private
    def article_params
      params.require(:article).permit(:title, :body, :author)
    end
# ...
```

Next, create a form page for article creation in `app/views/articles/new.html.erb`:

```ruby
<h1>New Article</h1>

<%= form_with model: @article do |form| %>
  <div>
    <%= form.label :title %><br>
    <%= form.text_field :title %>
  </div>

  <div>
    <%= form.label :body %><br>
    <%= form.text_area :body %>
  </div>

  <div>
    <%= form.label :author %><br>
    <%= form.text_field :author %>
  </div>

  <div>
    <%= form.submit %>
  </div>
<% end %>
```

> You should now have a nice Ruby on Rails app with a home page, article pages, and the ability to add new articles. Feel free to customize it more, but now it is time to add PostHog.

Reviewer: ```suggestion
You should now have a nice Ruby on Rails app with a home page, article pages, and the ability to add new articles. Customize it further if you'd like, then when you're ready it's time to add PostHog.
```

![Full app](../images/tutorials/ruby-on-rails-analytics/ruby.gif)

## 2. Integrating the PostHog snippet and Ruby library

> Now our Ruby on Rails app is ready to integrate with PostHog. There are two ways to do this through the snippet and the library. 

Reviewer: ```suggestion
Now our Ruby on Rails app is ready to integrate with PostHog. There are two ways to do this; with the snippet, or the library. The first method enables you to quickly get started with most PostHog features, such as analytics and session recording. If also want to use advanced tools, such as feature flags and experiments, you'll need to use the library. 
```

### Adding the JavaScript snippet

The first way is to use the JavaScript snippet. This provides autocapture of events, pageviews, session recordings, and more.

> To set it up, copy the snippet from getting started or your project settings and add it to `views/layouts/application.html.erb`. With no other changes, this file looks like this:

Reviewer: ```suggestion
To set it up, [copy the code snippet from the PostHog docs](/docs/integrate) and add it to `views/layouts/application.html.erb`. With no other changes, this file looks like this:
```

    <script>
        !function(t,e){var o,n,p,r;e.__SV||(window.posthog=e,e._i=[],e.init=function(i,s,a){function g(t,e){var o=e.split(".");2==o.length&&(t=t[o[0]],e=o[1]),t[e]=function(){t.push([e].concat(Array.prototype.slice.call(arguments,0)))}}(p=t.createElement("script")).type="text/javascript",p.async=!0,p.src=s.api_host+"/static/array.js",(r=t.getElementsByTagName("script")[0]).parentNode.insertBefore(p,r);var u=e;for(void 0!==a?u=e[a]=[]:a="posthog",u.people=u.people||[],u.toString=function(t){var e="posthog";return"posthog"!==a&&(e+="."+a),t||(e+=" (stub)"),e},u.people.toString=function(){return u.toString(1)+".people (stub)"},o="capture identify alias people.set people.set_once set_config register register_once unregister opt_out_capturing has_opted_out_capturing opt_in_capturing reset isFeatureEnabled onFeatureFlags".split(" "),n=0;n<o.length;n++)g(u,o[n]);e._i.push([i,s,a])},e.__SV=1)}(document,window.posthog||[]);
        posthog.init('<ph_project_api_key>',{api_host:'ph_instance_address})
    </script>
    <%= javascript_importmap_tags %>
  </head>

  <body>
    <%= yield %>
  </body>
</html>
```

> > Make sure to turn on session recordings in project settings by scrolling down to "Recordings" and toggling on "Record user sessions"

Reviewer: ```suggestion
> Make sure to turn on session recordings in your PostHog project settings by scrolling down to "Recordings" and toggling on "Record user sessions"
```

> Once you set this up and you click around on the server, you should see events in your PostHog instance from the source `web`.

Reviewer: ```suggestion
Once you've set this up and clicked around on the server, you should see events in your PostHog instance from the source `web`.
```

### Adding the Ruby library

If you want access to all the features of PostHog, including custom event capture, user identification, and feature flags, you need the Ruby library. 

To set this up, first, add PostHog to the `Gemfile`:

```ruby
gem 'posthog-ruby'
```

Next, in `config/initializers`, create a `posthog.rb` file and initialize PostHog as a global variable there like this:

```ruby
$posthog = PostHog::Client.new(
    api_key: '<ph_project_api_key>',
    host: '<ph_instance_address>',
    on_error: Proc.new { |status, msg| print msg }
)
```

Now, you can call `$posthog` throughout your Ruby on Rails application to access all the features of PostHog.

## 3. Using PostHog in your Ruby on Rails app

With both the library and snippet set up, we can use both of them to capture events, identify users, and use feature flags.

### Capturing custom events

> The snippet provides autocaptures of pageviews, clicks, inputs, and some other events. If you want customization, you must set up custom event capture.

Reviewer: ```suggestion
The snippet provides autocaptures of pageviews, clicks, inputs, and some other events. If you want to capture events which are more nuanced or unique to your product, you must set up custom event capture.
```
